In [ ]:
# install all libs
!pip install -r requirements.txt

# GPT4 with Retrieval Augmentation over Web

In this notebook we'll work through an example of using GPT-4 with retrieval augmentation to answer questions about NOUS Wissensmanagement GmbH.

In this example, we will download all the content from https://www.nousdigital.net/en/. 

This downloads all HTML into the `nous` directory.

In [ ]:
# crawl website content using wget
!wget -e robots=off -r -A.html -P nous https://www.nousdigital.net/en/

Now we can use LangChain itself to process these docs. 
We do this using the `DirectoryLoader` together with the `BSHTMLLoader:

In [8]:
from langchain.document_loaders import BSHTMLLoader, DirectoryLoader

loader = DirectoryLoader('nous/www.nousdigital.net/en', loader_cls=BSHTMLLoader)

docs = loader.load()
len(docs)

232

This leaves us with `232` processed doc pages. Let's take a look at the format each one contains:

In [72]:
docs[20].page_content

'                 NOUS solutions for culture       – NOUSdigital                       Data protection     This website uses cookies and services in the categories listed below. Select REQUIRED ONLY to accept only technically necessary cookies or SAVE CHOICE to save your individual settings.         FURTHER INFORMATION       Privacy Policy  Legal Notice                    Required Cookies       These cookies are needed to enable the basic functionality of this website. These cookies can therefore not be disabled.     HTTP Cookie: accepted_optional_cookies  Purpose: This cookie stores information about which optional cookies have been accepted or rejected. Domain: www.nousdigital.net Storage duration: 1 year  Third party: No      HTTP Cookie: csrftoken  Purpose: Protect against "Cross Site Request Forgery (CSRF)" attacks via form submission. Domain: www.nousdigital.net Storage duration: 1 year  Third party: No      HTTP Cookie: messages  Purpose: Stores messages from system so that they

In [46]:
#remove '\n' in docs.page_content
docs[20].page_content = docs[20].page_content.replace('\n', ' ')

Now let's see how we can process all of these. We will chunk everything into ~500 token chunks, we can do this easily with `langchain` and `tiktoken`:

In [2]:
import tiktoken

tokenizer_name = tiktoken.encoding_for_model('gpt-4')
tokenizer_name.name

'cl100k_base'

In [3]:
tokenizer = tiktoken.get_encoding(tokenizer_name.name)

# create the length function
def tiktoken_len(text):
    tokens = tokenizer.encode(
        text,
        disallowed_special=()
    )
    return len(tokens)

In [9]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=400,
    chunk_overlap=50,
    length_function=tiktoken_len,
    separators=["\n\n", "\n", " ", ""]
)

Process the `docs` into more chunks using this approach.

In [10]:
from typing_extensions import Concatenate
from uuid import uuid4
from tqdm.auto import tqdm

chunks = []

for idx, page in enumerate(tqdm(docs)):
    content = page.page_content
    if len(content) > 100:
        url = page.metadata['source'].replace('rtdocs/', 'https://')
        texts = text_splitter.split_text(content)
        chunks.extend([{
            'id': str(uuid4()),
            'text': texts[i],
            'chunk': i,
            'url': url
        } for i in range(len(texts))])

  0%|          | 0/232 [00:00<?, ?it/s]

100%|██████████| 232/232 [00:02<00:00, 100.91it/s]


In [11]:
chunks[2]

{'id': '0a2b7fc1-7e97-4d9d-a2b0-015d24096d4e',
 'text': 'Third party: No\n\n\n\n\n\nHTTP Cookie:\ncsrftoken\n\nPurpose: Protect against "Cross Site Request Forgery (CSRF)" attacks via form submission.\nDomain: www.nousdigital.net\nStorage duration: 1 year\n\nThird party: No\n\n\n\n\n\nHTTP Cookie:\nmessages\n\nPurpose: Stores messages from system so that they can be displayed for users.\nDomain: www.nousdigital.net\nStorage duration: Session\n\nThird party: No',
 'chunk': 2,
 'url': 'nous/www.nousdigital.net/en/index.html'}

Our chunks are ready so now we move onto embedding and indexing everything.

## Initialize Embedding Model

We use `text-embedding-ada-002` as the embedding model. We can embed text like so:

In [12]:
import os
import openai


openai.api_key = os.getenv("OPENAI_API_KEY") or "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"

openai.Engine.list()  # check we have authenticated

<OpenAIObject list at 0x7fb7e368eca0> JSON: {
  "object": "list",
  "data": [
    {
      "object": "engine",
      "id": "whisper-1",
      "ready": true,
      "owner": "openai-internal",
      "permissions": null,
      "created": null
    },
    {
      "object": "engine",
      "id": "babbage",
      "ready": true,
      "owner": "openai",
      "permissions": null,
      "created": null
    },
    {
      "object": "engine",
      "id": "davinci",
      "ready": true,
      "owner": "openai",
      "permissions": null,
      "created": null
    },
    {
      "object": "engine",
      "id": "text-davinci-edit-001",
      "ready": true,
      "owner": "openai",
      "permissions": null,
      "created": null
    },
    {
      "object": "engine",
      "id": "babbage-code-search-code",
      "ready": true,
      "owner": "openai-dev",
      "permissions": null,
      "created": null
    },
    {
      "object": "engine",
      "id": "text-similarity-babbage-001",
      "ready": t

In [13]:
embed_model = "text-embedding-ada-002"

res = openai.Embedding.create(
    input=[
        "Sample document text goes here",
        "there will be several phrases in each batch"
    ], engine=embed_model
)

In the response `res` we will find a JSON-like object containing our new embeddings within the `'data'` field.

In [14]:
res.keys()

dict_keys(['object', 'data', 'model', 'usage'])

Inside `'data'` we will find two records, one for each of the two sentences we just embedded. Each vector embedding contains `1536` dimensions (the output dimensionality of the `text-embedding-ada-002` model.

In [15]:
len(res['data'])

2

In [16]:
len(res['data'][0]['embedding']), len(res['data'][1]['embedding'])

(1536, 1536)

We will apply this same embedding logic to the langchain docs dataset we've just scraped. But before doing so we must create a place to store the embeddings.

## Initializing the Index

Now we need a place to store these embeddings and enable a efficient vector search through them all. To do that we use Pinecone, we can get a [free API key](https://app.pinecone.io/) and enter it below where we will initialize our connection to Pinecone and create a new index.

In [17]:
import pinecone

# setup pinecone account to get credentials
PINECONE_API_KEY='95c63bd8-3184-4d48-88bd-b08c41a94d84'
PINECONE_ENVIRONMENT='us-west1-gcp-free'

os.environ['PINECONE_ENVIRONMENT'] = PINECONE_ENVIRONMENT

pinecone.init(api_key=PINECONE_API_KEY, enviroment=PINECONE_ENVIRONMENT)
pinecone.whoami()

WhoAmIResponse(username='a9d81a4', user_label='default', projectname='cc97d47')

In [24]:
index_name = 'gpt-4-nous-docs'

In [25]:
import time

# check if index already exists (it shouldn't if this is first time)
if index_name not in pinecone.list_indexes():
    # if does not exist, create index
    pinecone.create_index(
        index_name,
        dimension=len(res['data'][0]['embedding']),
        metric='cosine'
    )
    # wait for index to be initialized
    time.sleep(1)

# connect to index
index = pinecone.Index(index_name)
# view index stats
index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0,
 'namespaces': {},
 'total_vector_count': 0}

We can see the index is currently empty with a `total_vector_count` of `0`. We can begin populating it with OpenAI `text-embedding-ada-002` built embeddings like so:

In [29]:
len(chunks)

4071

In [26]:
chunks[1]['text']

'Privacy Policy\n\nLegal Notice\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nRequired Cookies\n\n\n\n\n\n\nThese cookies are needed to enable the basic functionality of this website. These cookies can therefore not be disabled.\n\n\n\n\nHTTP Cookie:\naccepted_optional_cookies\n\nPurpose: This cookie stores information about which optional cookies have been accepted or rejected.\nDomain: www.nousdigital.net\nStorage duration: 1 year\n\nThird party: No\n\n\n\n\n\nHTTP Cookie:\ncsrftoken'

In [27]:
openai.Embedding.create(input=chunks[1]['text'], engine=embed_model)

<OpenAIObject list at 0x7fb7e49d0fe0> JSON: {
  "object": "list",
  "data": [
    {
      "object": "embedding",
      "index": 0,
      "embedding": [
        -0.0071838246658444405,
        -0.001452411524951458,
        0.017061583697795868,
        -0.03382384032011032,
        -0.03736133128404617,
        0.0004315566911827773,
        -0.003615722293034196,
        -0.010374368168413639,
        -0.005534130148589611,
        -0.03461297228932381,
        0.014027506113052368,
        -0.013333613984286785,
        0.0009830138878896832,
        -0.011476431973278522,
        -0.0010535935871303082,
        0.009911772795021534,
        0.021755559369921684,
        -0.03657219931483269,
        0.02412295714020729,
        -0.019592249765992165,
        -0.012973061762750149,
        0.015687404200434685,
        -0.03654498606920242,
        -0.00462254649028182,
        -0.006571566686034203,
        -0.019850758835673332,
        0.005962710361927748,
        -0.022735172882

In [28]:
from tqdm.auto import tqdm
from time import sleep

batch_size = 100  # how many embeddings we create and insert at once

for i in tqdm(range(0, len(chunks), batch_size)):
    # find end of batch
    i_end = min(len(chunks), i+batch_size)
    meta_batch = chunks[i:i_end]
    # get ids
    ids_batch = [x['id'] for x in meta_batch]
    # get texts to encode
    texts = [x['text'] for x in meta_batch]
    # create embeddings (try-except added to avoid RateLimitError)
    try:
        res = openai.Embedding.create(input=texts, engine=embed_model)
    except Exception as e:
        print(e)
        done = False
        while not done:
            sleep(5)
            try:
                res = openai.Embedding.create(input=texts, engine=embed_model)
                done = True
            except:
                pass
    embeds = [record['embedding'] for record in res['data']]
    # cleanup metadata
    meta_batch = [{
        'text': x['text'],
        'chunk': x['chunk'],
        'url': x['url']
    } for x in meta_batch]
    to_upsert = list(zip(ids_batch, embeds, meta_batch))
    # upsert to Pinecone
    index.upsert(vectors=to_upsert)

100%|██████████| 41/41 [01:41<00:00,  2.48s/it]


Now we've added all of our langchain docs to the index. With that we can move on to retrieval and then answer generation using GPT-4.

## Retrieval

To search through our documents we first need to create a query vector `xq`. Using `xq` we will retrieve the most relevant chunks from the LangChain docs, like so:

In [30]:
def create_embedding(query):
    return openai.Embedding.create(
        input=[query],
        engine=embed_model
    )

def query_pinecone_embedding(openai_embedding):
    # retrieve from Pinecone
    xq = openai_embedding['data'][0]['embedding']
    # get relevant contexts (including the questions)
    return index.query(xq, top_k=5, include_metadata=True)

In [31]:
query = "Who is the CEO of NOUS Wissensmanagement GmbH?"

embedding = create_embedding(query)
result = query_pinecone_embedding(embedding)
result

{'matches': [{'id': 'dde41bbd-08b6-425a-867c-89c7266e4e2f',
              'metadata': {'chunk': 17.0,
                           'text': 'NOUS Wissensmanagement GmbH\n'
                                   '\n'
                                   '          Ullmannstraße 35 |\n'
                                   '          1150 Vienna, Austria |\n'
                                   '          info@nousdigital.com\n'
                                   '\n'
                                   '\n'
                                   '\n'
                                   'Newsletter signup\n'
                                   '\n'
                                   '          register\n'
                                   '        \n'
                                   '\n'
                                   '\n'
                                   '\n'
                                   '\n'
                                   '\n'
                                   '\n'
                  

With retrieval complete, we move on to feeding these into GPT-4 to produce answers.

## Retrieval Augmented Generation

GPT-4 is currently accessed via the `ChatCompletions` endpoint of OpenAI. To add the information we retrieved into the model, we need to pass it into our user prompts *alongside* our original query. We can do that like so:

In [32]:
def pinecone_to_list(res):
    # get list of retrieved texts
    contexts = [item['metadata']['text'] for item in res['matches']]
    return "\n\n---\n\n".join(contexts)+"\n\n-----\n\n"+query

pinecone_to_list(result)


'NOUS Wissensmanagement GmbH\n\n          Ullmannstraße 35 |\n          1150 Vienna, Austria |\n          info@nousdigital.com\n\n\n\nNewsletter signup\n\n          register\n        \n\n\n\n\n\n\n\n                  Home\n                \n\n\n\n                  What We Do\n                \n\n\n\n                  Our Products\n                \n\n\n\n                  Our Showroom\n                \n\n\n\n                  About us\n                \n\n\n\n                  Blog\n                \n\n\n\n\nLegal Notice\n\n\nTerms & Conditions\n\n---\n\nNOUS Wissensmanagement GmbH\n\n          Ullmannstraße 35 |\n          1150 Vienna, Austria |\n          info@nousdigital.com\n\n\n\nNewsletter signup\n\n          register\n        \n\n\n\n\n\n\n\n                  Home\n                \n\n\n\n                  What We Do\n                \n\n\n\n                  Our Products\n                \n\n\n\n                  Our Showroom\n                \n\n\n\n                  About us

Now we configure following prompt:

In [33]:
def query_gpt(prompt, message, model="gpt-4"):
     return openai.ChatCompletion.create(
        model=model,
        messages=[
            {"role": "system", "content": prompt},
            {"role": "user", "content": message}
        ]
    )

In [40]:
# system message to 'prime' the model
prompt = f"""You are Q&A bot. A highly intelligent system that answers
user questions based on the information provided by the user above
each question. If the information can not be found in the information
provided by the user you truthfully say "I don't know".
"""

In [41]:
def search_docs(query, model="gpt-4"):
    embedding = create_embedding(query)
    result = query_pinecone_embedding(embedding)
    open_ai_message = pinecone_to_list(result)
    return query_gpt(prompt, open_ai_message, model=model)

In [45]:
query = "What services does NOUS provide?"
# "What can you tell me about the company NOUS?"
# "What services does NOUS provide?"
# "Who are the CEOs?"
# "Who runs the company?"
# "Who is the CEO of NOUS Wissensmanagement?"
# "How does iteratec talk about digital champions?
# "What services does NOUS Wissensmanagement provide?"
# "Which technologies does iteratec use?"

result = search_docs(query, model="gpt-4")
result

<OpenAIObject chat.completion id=chatcmpl-7QxmRpFpENcGNWfsNhmrCvyuuH4an at 0x7fb7e25d6ca0> JSON: {
  "id": "chatcmpl-7QxmRpFpENcGNWfsNhmrCvyuuH4an",
  "object": "chat.completion",
  "created": 1686660679,
  "model": "gpt-4-0314",
  "usage": {
    "prompt_tokens": 388,
    "completion_tokens": 31,
    "total_tokens": 419
  },
  "choices": [
    {
      "message": {
        "role": "assistant",
        "content": "NOUS provides services in the field of mobile guides and media-based education. They create tailor-made concepts with innovative technology and specialize in developing solutions for culture."
      },
      "finish_reason": "stop",
      "index": 0
    }
  ]
}

In [56]:
result

<OpenAIObject chat.completion id=chatcmpl-7QkUNvMEJl9Tl5mybpybz5Zx2CXio at 0x7f863c152070> JSON: {
  "id": "chatcmpl-7QkUNvMEJl9Tl5mybpybz5Zx2CXio",
  "object": "chat.completion",
  "created": 1686609587,
  "model": "gpt-4-0314",
  "usage": {
    "prompt_tokens": 571,
    "completion_tokens": 16,
    "total_tokens": 587
  },
  "choices": [
    {
      "message": {
        "role": "assistant",
        "content": "The CEO of NOUS Wissensmanagement GmbH is Raphael Schneeberger."
      },
      "finish_reason": "stop",
      "index": 0
    }
  ]
}

To display this response nicely, we will display it in markdown.

In [128]:
from IPython.display import Markdown

display(Markdown(result['choices'][0]['message']['content']))

NOUS is a multinational company that specializes in mobile guides and media-based education. Headquartered in Vienna, Austria, it has branch offices in Denver (USA) and Dubai (UAE). NOUS has worked on over 200 projects with museums, cultural organizations, and big brands worldwide. They advise institutions and companies on adapting their educational roles to align with the digital world and develop multimedia, intuitive end-user, and streaming applications on the web and other platforms.

Let's compare this to a non-augmented query...

In [35]:
res = openai.ChatCompletion.create(
    model="gpt-4",
    messages=[
        {"role": "system", "content": prompt },
        {"role": "user", "content": query}
    ]
)
display(Markdown(res['choices'][0]['message']['content']))

I don't know.

If we drop the `"I don't know"` part of the `primer`?

In [36]:
res = openai.ChatCompletion.create(
    model="gpt-4",
    messages=[
        {"role": "system", "content": "You are Q&A bot. A highly intelligent system that answers user questions"},
        {"role": "user", "content": query}
    ]
)
display(Markdown(res['choices'][0]['message']['content']))

As of my last knowledge update in September 2021, the CEO of iteratec is Michael Ernst. However, it's worth checking the most recent information on the company's website or other sources, as management positions might have changed.

Then we see something even worse than `"I don't know"` — hallucinations. Clearly augmenting our queries with additional context can make a huge difference to the performance of our system.

Great, we've seen how to augment GPT-4 with semantic search to allow us to answer LangChain specific queries.

Once you're finished, we delete the index to save resources.

In [19]:
#pinecone.delete_index(index_name)